docs(user-flows): refactor document (#14021)
GoogleChrome/lighthouse: docs/user-flows.md

# User Flows in Lighthouse

Historically, Lighthouse has analyzed the cold pageload of a page. Starting in 2022 (Lighthouse v10), it can analyze and report on the entire page lifecycle via "user flows".

#### You might be interested in flows if…

* … you want to run Lighthouse on your whole webapp, not just the landing page.

> * … you want to know if the checkout process is accessible.

Reviewer: ```suggestion
* … you want to know if all parts of the page are accessible (e.g. a checkout process).
```

* … you want to know the Cumulative Layout Shift of my SPA page transition

In these cases, you want Lighthouse on a _flow_, not just a page load.

## The three modes: Navigation, Timespan, Snapshot

Lighthouse can now run in three modes: navigations, timespans, and snapshots. Each mode has its own unique use cases, benefits, and limitations. Later, you'll create a flow by combining these three core report types.

> * **Navigation mode** analyzes a single page load. Prior to v10, all Lighthouse runs were essentially in this mode.

Reviewer: Not sure if we should be using v10 as the time origin since the modes were released with v9.6.1

* **Timespan mode** analyzes an arbitrary period of time, typically containing user interactions.
* **Snapshot mode** analyzes the page in a particular state.

| | |
|:---:|---|

> | Navigation <br> <img src="https://user-images.githubusercontent.com/39191/168928225-f2157fda-5131-4bd0-9121-b1a0b2f869a7.png" height="96" width="141">  | **Benefits** <br>  ✅ Provides an overall performance score and all metrics.<br>✅ Contains the most advice of all report types.<br><br> **Limitations** <br>  🤔 Cannot analyze form submissions or single page app transitions.<br>🤔 Cannot analyze content that isn't available immediately on page load.<br><br> **Use Cases** <br>  👍 Obtain a Lighthouse Performance score.<br>👍 Measure all performance metrics.<br>👍 Assess Progressive Web App capabilities. |

Reviewer: Could you also add these images to the more detailed sections below this table so there is a god visual callout. I was thinking something similar to the "Creating a Flow" section.

Author: good idea. done

> | Navigation <br> <img src="https://user-images.githubusercontent.com/39191/168928225-f2157fda-5131-4bd0-9121-b1a0b2f869a7.png" height="96" width="141">  | **Benefits** <br>  ✅ Provides an overall performance score and all metrics.<br>✅ Contains the most advice of all report types.<br><br> **Limitations** <br>  🤔 Cannot analyze form submissions or single page app transitions.<br>🤔 Cannot analyze content that isn't available immediately on page load.<br><br> **Use Cases** <br>  👍 Obtain a Lighthouse Performance score.<br>👍 Measure all performance metrics.<br>👍 Assess Progressive Web App capabilities. |

Reviewer: I'm not sure we need a "Benefits" *and* and "Use Cases" section for each mode. A lot of the points are repeated between the two sections. WDYT about just having a "Use Cases" and "Limitations" section?

| Timespan <br> <img src="https://user-images.githubusercontent.com/39191/168928251-c7025cd5-0086-4db8-ae52-95a5b5675adf.png" height="96" width="141">  | **Benefits** <br>  ✅ Provides timerange-based metrics such as TBT, CLS and INP.<br>✅ Analyzes any period of time, including user interactions or single page app transitions.<br><br> **Limitations** <br>  🤔 Does not provide an overall performance score.<br>🤔 Cannot analyze moment-based performance metrics (e.g. Largest Contentful Paint).<br>🤔 Cannot analyze state-of-the-page issues (e.g. no Accessibility category)<br><br> **Use Cases** <br>  👍 Measure layout shifts and JavaScript execution time on a series of interactions.<br>👍 Discover performance opportunities to improve the experience for long-lived pages and SPAs. |

> | Snapshot <br> <img src="https://user-images.githubusercontent.com/39191/168931653-b45e0b6b-c5bd-4d8d-85b6-fee8425186a0.png" height="96" width="141">  | **Benefits** <br>   ✅ Analyzes the page in its current state.<br><br> **Limitations** <br>  🤔 Does not provide an overall performance score or metrics.<br>🤔 Cannot analyze any issues outside the current DOM (e.g. no network, main-thread, or performance analysis).<br><br> **Use Cases** <br>  👍 Find accessibility issues in single page applications or complex forms.<br>👍 Evaluate best practices of menus and UI elements hidden behind interaction. |

Reviewer: This is kinda hard to read/review from the md side. Is it possible to use newlines to separate each bullet point?

Author: unfortunately i can't. but the text is the same as what it was. (except i shortened a few metric references)

  await page.goto('https://secret.login');
  const flow = await lighthouse.startFlow(page);

  await flow.beginTimespan();
  await page.type('#password', 'L1ghth0useR0cks!');
  await page.click('#login');
  await page.waitForSelector('#dashboard');
  await flow.endTimespan();

  await browser.close();
  writeFileSync('report.html', await flow.generateReport());
})();
```

### Snapshot

In DevTools, select "Snapshot" as the mode. Set up the page in the state you want to evaluate. Then, click _Analyze page state_.

![snapshotdt](https://user-images.githubusercontent.com/39191/168929172-92a70108-a053-4dda-b719-2900b9d3d956.png)

#### Snapshots in the Node.js API

```js
import {writeFileSync} from 'fs';
import puppeteer from 'puppeteer';
import lighthouse from 'lighthouse/lighthouse-core/fraggle-rock/api.js';

(async function() {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  await page.goto('https://example.com');
  const flow = await lighthouse.startFlow(page);

  await page.click('#expand-sidebar');
  await flow.snapshot();

  await browser.close();
  writeFileSync('report.html', await flow.generateReport());
})();
```

## Creating a Flow

> <img src="https://user-images.githubusercontent.com/39191/168931109-5ae80620-7464-4d98-a834-71169a5b6a61.png" height=300>

Reviewer: Can this guy be a little smaller, maybe 200px?

Author: matches the rest now. 240px

# User Flows in Lighthouse

Historically, Lighthouse has analyzed the cold pageload of a page. Starting in 2022 (Lighthouse v10), it can analyze and report on the entire page lifecycle via "user flows".

#### You might be interested in flows if…

* … you want to run Lighthouse on your whole webapp, not just the landing page.
* … you want to know if the checkout process is accessible.
* … you want to know the Cumulative Layout Shift of my SPA page transition

In these cases, you want Lighthouse on a _flow_, not just a page load.

## The three modes: Navigation, Timespan, Snapshot

Lighthouse can now run in three modes: navigations, timespans, and snapshots. Each mode has its own unique use cases, benefits, and limitations. Later, you'll create a flow by combining these three core report types.

> * <img src="https://user-images.githubusercontent.com/39191/168928225-f2157fda-5131-4bd0-9121-b1a0b2f869a7.png" height="80" align="middle"> **Navigation mode** analyzes a single page load. Prior to v10, all Lighthouse runs were essentially in this mode.

Reviewer: IMO these images would be better to visually identify the headings below. I know I'm in the user flow section cuz I saw the user flow image, so maybe that's how these images should be used.

I also think they look kinda weird as part of a bulleted list.

Author: Yeah. you'll see in that screenshot above i experimented with imagery in the table.
agree the inline image on bullet points is kinda whack

ill try this out.. i think it'll work out well. :)

* <img src="https://user-images.githubusercontent.com/39191/168928251-c7025cd5-0086-4db8-ae52-95a5b5675adf.png" height="80" align="middle"> **Timespan mode** analyzes an arbitrary period of time, typically containing user interactions.
* <img src="https://user-images.githubusercontent.com/39191/168931653-b45e0b6b-c5bd-4d8d-85b6-fee8425186a0.png" height="80" align="middle"> **Snapshot mode** analyzes the page in a particular state.

> |  |  |

Reviewer: ```suggestion
```

|---|---|

> | Navigation | <small> *Benefits* <br> ✅ Provides an overall performance score and all metrics.<br>✅ Contains the most advice of all report types (both time-based and state-based audits are available).<br> *Limitations* <br> 🤔 Cannot analyze form submissions or single page app transitions.<br>🤔 Cannot analyze content that isn't available immediately on page load.<br> *Use Cases* <br> 👍 Obtain a Lighthouse Performance score.<br>👍 Measure Performance metrics (First Contentful Paint, Largest Contentful Paint, Speed Index, Time to Interactive, Cumulative Layout Shift, Total Blocking Time).<br>👍 Assess Progressive Web App capabilities.</small> |

Reviewer: If we're going to use a table here, can benefits/limitations/use cases each have there own column? If that doesn't look good then maybe we're better off using headings/subheadings for this.

Author: i tried the per-column look but there's just too much line-wrapping and it looked a lil whack.

(well.. what i had was slightly diff but the wrapping is similar to what it'd be:)
![image](https://user-images.githubusercontent.com/39191/169391787-365ff1a3-fb48-47f7-9fc2-c0fda38ecb2b.png)

Lemme try some clearer headings…

Author: yup better now
![image](https://user-images.githubusercontent.com/39191/169396931-1537d5e4-8056-414b-b017-39f39321ae10.png)

| Snapshot | <small> *Benefits* <br>  ✅ Analyzes the page in its current state.<br> *Limitations* <br> 🤔 Does not provide an overall performance score or metrics.<br>🤔 Cannot analyze any issues outside the current DOM (e.g. no network, main-thread, or performance analysis).<br> *Use Cases* <br> 👍 Find accessibility issues in single page applications or complex forms.<br>👍 Evaluate best practices of menus and UI elements hidden behind interaction. |

### Navigation mode

In DevTools, navigation is easy: ensure it's the selected mode and then click _Analyze page load_.

![navdt](https://user-images.githubusercontent.com/39191/168929174-11311144-ce9b-4124-9a52-0423a073b9fe.png)

> Note: DevTools only generates a report for a standalone navigation. To combine it with other steps for a multi-step user flow report, [use the Node API](#creating-a-flow).

#### Navigations in the Node.js API

```js
import {writeFileSync} from 'fs';
import puppeteer from 'puppeteer';

> import lighthouse from 'lighthouse/lighthouse-core/fraggle-rock/api.js';

Author: yeah that makes sense. but i also think `api` is too general. it looks 'fine' on this line but then later with `api.startFlow` it's lost its identity

`lighthouseapi` ? `lhapi`?  

we should also fix the export so that `import lighthouse from 'lighthouse'` actually would work for all this. i feel good about doing that for v10

> import lighthouse from 'lighthouse/lighthouse-core/fraggle-rock/api.js';

Reviewer: > we should also fix the export so that import lighthouse from 'lighthouse' actually would work for all this. i feel good about doing that for v10

Ok, in that case we can leave this with plans to update the docs once `startFlow` is exported via the default entry point.

# User Flows in Lighthouse

Historically, Lighthouse has analyzed the cold pageload of a page. Starting in 2022 (Lighthouse v10), it can analyze and report on the entire page lifecycle via "user flows".

#### You might be interested in flows if…

* … you want to run Lighthouse on your whole webapp, not just the landing page.
* … you want to know if all parts of the web experience are accessible (e.g. a checkout process).
* … you want to know the Cumulative Layout Shift of my SPA page transition

In these cases, you want Lighthouse on a _flow_, not just a page load.

## The three modes: Navigation, Timespan, Snapshot

Lighthouse can now run in three modes: navigations, timespans, and snapshots. Each mode has its own unique use cases, benefits, and limitations. Later, you'll create a flow by combining these three core report types.

> * **Navigation mode** analyzes a single page load. Prior to v9.6.1, all Lighthouse runs were essentially in this mode.

Reviewer: I think 9.6 makes more sense here
```suggestion
* **Navigation mode** analyzes a single page load. Prior to v9.6.0, all Lighthouse runs were essentially in this mode.
```

Historically, Lighthouse has analyzed the cold pageload of a page. Starting in 2022 (Lighthouse v10), it can analyze and report on the entire page lifecycle via "user flows".

#### You might be interested in flows if…

* … you want to run Lighthouse on your whole webapp, not just the landing page.
* … you want to know if all parts of the web experience are accessible (e.g. a checkout process).
* … you want to know the Cumulative Layout Shift of my SPA page transition

In these cases, you want Lighthouse on a _flow_, not just a page load.

## The three modes: Navigation, Timespan, Snapshot

Lighthouse can now run in three modes: navigations, timespans, and snapshots. Each mode has its own unique use cases, benefits, and limitations. Later, you'll create a flow by combining these three core report types.

* **Navigation mode** analyzes a single page load. It should feel familiar as all Lighthouse runs prior to v9.6.0 were essentially in this mode. Even in v9.6.0 and onwards, it remains the default.
* **Timespan mode** analyzes an arbitrary period of time, typically containing user interactions.
* **Snapshot mode** analyzes the page in a particular state.

| | |
|:---:|---|

> | Navigation <br> <img src="https://user-images.githubusercontent.com/39191/170560057-4f7cab97-ee6b-497f-8e18-1f84ba3616dd.png" height="96" width="141">  | **Use Cases** <br>  ✅ Obtain a Lighthouse Performance score and all performance metrics.<br>✅ Assess Progressive Web App capabilities.<br><br> **Limitations** <br>  🤔 Cannot analyze form submissions or single page app transitions.<br>🤔 Cannot analyze content that isn't available immediately on page load. |

Author: done

> | Timespan <br> <img src="https://user-images.githubusercontent.com/39191/170560049-a6b630a2-187a-4233-9f8b-8d7fb276f51e.png" height="96" width="141">  | **Use Cases** <br> ✅ Measure layout shifts and JavaScript execution time over a timerange including interactions.<br>✅ Discover performance opportunities to improve the experience for long-lived pages and SPAs. <br>✅ Quantify single page app transitions.<br><br> **Limitations** <br>  🤔 Does not provide an overall performance score.<br>🤔 Cannot analyze moment-based performance metrics (e.g. Largest Contentful Paint).<br>🤔 Cannot analyze state-of-the-page issues (e.g. no Accessibility category) |

Reviewer: I don't think we need "Quantify single page app transitions" because of the bullet point above it

Author: done

Lighthouse can now run in three modes: navigations, timespans, and snapshots. Each mode has its own unique use cases, benefits, and limitations. Later, you'll create a flow by combining these three core report types.

> * **Navigation mode** analyzes a single page load. Prior to v9.6.0, all Lighthouse runs were essentially in this mode.

Reviewer 2: Reword? I can't tell at a glance without employing multiple brain cells if this is exclusive or not :')

Author: how about

> * **Navigation mode** analyzes a single page load. It should feel familiar as all Lighthouse runs prior to v9.6.0 were essentially in this mode. Even in v9.6.0 and onwards, it remains the default.
